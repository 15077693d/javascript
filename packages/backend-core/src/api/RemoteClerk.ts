import {
  AllowlistIdentifierApi,
  ClientApi,
  EmailApi,
  InvitationApi,
  SessionApi,
  SMSMessageApi,
  UserApi,
} from './collection';
import RestClient, { ClerkFetcher } from './utils/RestClient';

const defaultApiKey = process.env.CLERK_API_KEY || '';
const defaultApiVersion = process.env.CLERK_API_VERSION || 'v1';
const defaultServerApiUrl =
  process.env.CLERK_API_URL || 'https://api.clerk.dev';

type RemoteClerkProps = {
  apiKey?: string;
  serverApiUrl?: string;
  apiVersion?: string;
  fetcher: ClerkFetcher;
  libName: string;
  libVersion: string;
  packageRepo: string;
};

export class RemoteClerk {
  private _restClient: RestClient;
  // singleton instance
  static _instance: RemoteClerk;

  // TODO we may not need to instantiate these any more if they keep no state
  // private api instances
  private _allowlistIdentifierApi?: AllowlistIdentifierApi;
  private _clientApi?: ClientApi;
  private _emailApi?: EmailApi;
  private _invitationApi?: InvitationApi;
  private _sessionApi?: SessionApi;
  private _smsMessageApi?: SMSMessageApi;
  private _userApi?: UserApi;

  constructor({
    apiKey = defaultApiKey,
    serverApiUrl = defaultServerApiUrl,
    apiVersion = defaultApiVersion,
    fetcher,
    libName,
    libVersion,
    packageRepo,
  }: RemoteClerkProps) {
    //   if (!apiKey) {
    //     throw Error(SupportMessages.API_KEY_NOT_FOUND);
    //   }

    this._restClient = new RestClient(
      apiKey,
      serverApiUrl,
      apiVersion,
      fetcher,
      libName,
      libVersion,
      packageRepo
    );
  }

  fetchInterstitial<T>(): Promise<T> {
    return this._restClient.fetchInterstitial<T>();
  }

  // // For use as singleton, always returns the same instance
  // static getInstance(): RemoteClerk {
  //   if (!this._instance) {
  //     this._instance = new RemoteClerk();
  //   }

  //   return this._instance;
  // }

  // Setters for the embedded rest client
  set apiKey(value: string) {
    this._restClient.apiKey = value;
  }

  set serverApiUrl(value: string) {
    this._restClient.serverApiUrl = value;
  }

  set apiVersion(value: string) {
    this._restClient.apiVersion = value;
  }

  // Lazy sub-api getters
  get allowlistIdentifiers(): AllowlistIdentifierApi {
    if (!this._allowlistIdentifierApi) {
      this._allowlistIdentifierApi = new AllowlistIdentifierApi(
        this._restClient
      );
    }
    return this._allowlistIdentifierApi;
  }

  get clients(): ClientApi {
    if (!this._clientApi) {
      this._clientApi = new ClientApi(this._restClient);
    }

    return this._clientApi;
  }

  get emails(): EmailApi {
    if (!this._emailApi) {
      this._emailApi = new EmailApi(this._restClient);
    }

    return this._emailApi;
  }

  get invitations(): InvitationApi {
    if (!this._invitationApi) {
      this._invitationApi = new InvitationApi(this._restClient);
    }
    return this._invitationApi;
  }

  get sessions(): SessionApi {
    if (!this._sessionApi) {
      this._sessionApi = new SessionApi(this._restClient);
    }

    return this._sessionApi;
  }

  get smsMessages(): SMSMessageApi {
    if (!this._smsMessageApi) {
      this._smsMessageApi = new SMSMessageApi(this._restClient);
    }

    return this._smsMessageApi;
  }

  get users(): UserApi {
    if (!this._userApi) {
      this._userApi = new UserApi(this._restClient);
    }

    return this._userApi;
  }
}
