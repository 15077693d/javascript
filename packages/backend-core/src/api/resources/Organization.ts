import camelcaseKeys from 'camelcase-keys';

import filterKeys from '../utils/Filter';
import type { OrganizationJSON } from './JSON';
import type { OrganizationProps } from './Props';

interface OrganizationPayload extends OrganizationProps {}

export interface Organization extends OrganizationPayload {}

export class Organization {
  static attributes = ['id', 'name', 'slug', 'publicMetadata', 'privateMetadata', 'createdAt', 'updatedAt'];

  static defaults = [];

  constructor(data: Partial<OrganizationProps> = {}) {
    Object.assign(this, Organization.defaults, data);
  }

  static fromJSON(data: OrganizationJSON): Organization {
    const camelcased = camelcaseKeys(data);
    const filtered = filterKeys(camelcased, Organization.attributes);
    return new Organization(filtered as OrganizationPayload);
  }
}
