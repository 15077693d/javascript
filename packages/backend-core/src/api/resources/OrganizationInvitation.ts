import camelcaseKeys from 'camelcase-keys';

import filterKeys from '../utils/Filter';
import type { OrganizationInvitationJSON } from './JSON';
import type { OrganizationInvitationProps } from './Props';

export interface OrganizationInvitation extends OrganizationInvitationProps {}

export class OrganizationInvitation {
  static attributes = [
    'id',
    'emailAddress',
    'organizationId',
    'role',
    'redirectUrl',
    'status',
    'createdAt',
    'updatedAt',
  ];

  static defaults = [];

  constructor(data: OrganizationInvitationProps) {
    Object.assign(this, OrganizationInvitation.defaults, data);
  }

  static fromJSON(data: OrganizationInvitationJSON): OrganizationInvitation {
    const camelcased = camelcaseKeys(data);
    const filtered = filterKeys(camelcased, OrganizationInvitation.attributes);
    return new OrganizationInvitation(filtered as OrganizationInvitationProps);
  }
}
