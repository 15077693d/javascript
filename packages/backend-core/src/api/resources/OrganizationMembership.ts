import camelcaseKeys from 'camelcase-keys';

import { Organization } from '../resources';
import filterKeys from '../utils/Filter';
import type { OrganizationMembershipJSON, OrganizationMembershipPublicUserDataJSON } from './JSON';
import type { OrganizationMembershipProps, OrganizationMembershipPublicUserDataProps } from './Props';

export interface OrganizationMembership extends OrganizationMembershipProps {
  organization: Organization;
  publicUserData: OrganizationMembershipPublicUserData;
}

export class OrganizationMembership {
  static attributes = ['id', 'role', 'organization', 'publicUserData', 'createdAt', 'updatedAt'];

  static defaults = [];

  constructor(data: OrganizationMembershipProps) {
    Object.assign(this, OrganizationMembership.defaults, data);
  }

  static fromJSON(data: OrganizationMembershipJSON): OrganizationMembership {
    const camelcased = camelcaseKeys(data);
    const filtered = filterKeys(camelcased, OrganizationMembership.attributes);
    filtered.organization = Organization.fromJSON(data.organization);
    filtered.publicUserData = OrganizationMembershipPublicUserData.fromJSON(data.public_user_data);
    return new OrganizationMembership(filtered as OrganizationMembershipProps);
  }
}

export interface OrganizationMembershipPublicUserData extends OrganizationMembershipPublicUserDataProps {}

export class OrganizationMembershipPublicUserData {
  static attributes = ['identifier', 'firstName', 'lastName', 'profileImageUrl', 'userId'];
  static defaults = [];

  constructor(data: Partial<OrganizationMembershipPublicUserDataProps> = {}) {
    Object.assign(this, OrganizationMembershipPublicUserData.defaults, data);
  }

  static fromJSON(data: OrganizationMembershipPublicUserDataJSON): OrganizationMembershipPublicUserData {
    const camelcased = camelcaseKeys(data);
    const filtered = filterKeys(camelcased, OrganizationMembershipPublicUserData.attributes);
    return new OrganizationMembershipPublicUserData(filtered as OrganizationMembershipPublicUserDataProps);
  }
}
