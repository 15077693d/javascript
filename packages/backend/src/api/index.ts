export * from './factory';
export * from './HttpTransport';
