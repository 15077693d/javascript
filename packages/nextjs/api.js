let exportLib;
if (process.env.NEXT_RUNTIME === 'edge') {
  exportLib = require('./dist/edge-middleware');
  exportLib.withAuth = exportLib.withEdgeMiddlewareAuth;
} else {
  exportLib = require('./dist/api');
}

module.exports = exportLib;
