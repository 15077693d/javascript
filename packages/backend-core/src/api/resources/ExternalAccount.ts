import associationDefaults from '../utils/Associations';
import { Association } from './Enums';
import type { ExternalAccountJSON } from './JSON';
import { ObjectType } from './JSON';
import type { ExternalAccountProps } from './Props';
import { userAssociations } from './User';
import { Verification } from './Verification';

interface ExternalAccountPayload extends ExternalAccountProps {}

export interface ExternalAccount extends ExternalAccountPayload {}

export class ExternalAccount {
  static attributes = [
    'id',
    'provider',
    'externalId',
    'approvedScopes',
    'emailAddress',
    'firstName',
    'lastName',
    'picture',
    'username',
    'publicMetadata',
    'label',
  ];

  static associations = {
    verification: Association.HasOne,
  };

  static defaults = {
    ...associationDefaults(userAssociations),
  };

  constructor(data: Partial<ExternalAccountPayload> = {}) {
    Object.assign(this, ExternalAccount.defaults, data);
  }

  static fromJSON(data: ExternalAccountJSON): ExternalAccount {
    const obj = {} as ExternalAccountPayload;

    obj.id = data.id;
    obj.approvedScopes = data.approved_scopes;
    obj.emailAddress = data.email_address;
    obj.username = data.username;
    obj.publicMetadata = data.public_metadata;
    obj.label = data.label;

    if (data.verification) {
      obj.verification = Verification.fromJSON(data.verification);
    }

    switch (data.object) {
      case ObjectType.FacebookAccount: {
        obj.provider = 'facebook';
        const fbData = data;
        obj.externalId = fbData.facebook_id;
        obj.firstName = fbData.first_name;
        obj.lastName = fbData.last_name;
        obj.picture = data.picture;
        break;
      }
      case ObjectType.GoogleAccount: {
        obj.provider = 'google';
        const gData = data;
        obj.externalId = gData.google_id;
        obj.firstName = gData.given_name;
        obj.lastName = gData.family_name;
        obj.picture = data.picture;
        break;
      }
      case ObjectType.ExternalAccount: {
        obj.provider = data.provider;
        const extData = data;
        obj.externalId = extData.provider_user_id;
        obj.firstName = extData.first_name;
        obj.lastName = extData.last_name;
        obj.picture = extData.avatar_url;
        break;
      }
      default:
        throw new Error('Unsupported external account type');
    }

    return new ExternalAccount(obj);
  }
}
