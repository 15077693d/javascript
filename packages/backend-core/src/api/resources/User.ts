import camelcaseKeys from 'camelcase-keys';

import { EmailAddress } from './EmailAddress';
import { Association } from './Enums';
import { ExternalAccount } from './ExternalAccount';
import type { ExternalAccountJSON, UserJSON } from './JSON';
import { PhoneNumber } from './PhoneNumber';
import type { EmailAddressProps, ExternalAccountProps, PhoneNumberProps, Web3WalletProps } from './Props';
import { Web3Wallet } from './Web3Wallet';

export const userAssociations = {
  emailAddresses: Association.HasMany,
  phoneNumbers: Association.HasMany,
  web3Wallets: Association.HasMany,
  externalAccounts: Association.HasMany,
};

export class User {
  constructor(
    readonly id: string,
    readonly passwordEnabled: boolean,
    readonly twoFactorEnabled: boolean,
    readonly createdAt: number,
    readonly updatedAt: number,
    readonly profileImageUrl: string,
    readonly gender: string,
    readonly birthday: string,
    readonly primaryEmailAddressId: string | null,
    readonly primaryPhoneNumberId: string | null,
    readonly primaryWeb3WalletId: string | null,
    readonly lastSignInAt: string | null,
    readonly externalId: string | null,
    readonly username: string | null,
    readonly firstName: string | null,
    readonly lastName: string | null,
    readonly publicMetadata: Record<string, unknown> = {},
    readonly privateMetadata: Record<string, unknown> = {},
    readonly unsafeMetadata: Record<string, unknown> = {},
    readonly emailAddresses: EmailAddressProps[] = [],
    readonly phoneNumbers: PhoneNumberProps[] = [],
    readonly web3Wallets: Web3WalletProps[] = [],
    readonly externalAccounts: ExternalAccountProps[] = [],
  ) {}

  static fromJSON(data: UserJSON): User {
    const obj = camelcaseKeys(data) as Record<string, any>;

    obj.emailAddresses = (data.email_addresses || []).map(x => EmailAddress.fromJSON(x));
    obj.phoneNumbers = (data.phone_numbers || []).map(x => PhoneNumber.fromJSON(x));
    obj.web3Wallets = (data.web3_wallets || []).map(x => Web3Wallet.fromJSON(x));
    obj.externalAccounts = (data.external_accounts || []).map((x: ExternalAccountJSON) => ExternalAccount.fromJSON(x));

    return new User(
      obj.id,
      obj.passwordEnabled,
      obj.twoFactorEnabled,
      obj.createdAt,
      obj.updatedAt,
      obj.profileImageUrl,
      obj.gender,
      obj.birthday,
      obj.primaryEmailAddressId,
      obj.primaryPhoneNumberId,
      obj.primaryWeb3WalletId,
      obj.lastSignInAt,
      obj.externalId,
      obj.username,
      obj.firstName,
      obj.lastName,
      obj.publicMetadata,
      obj.privateMetadata,
      obj.unsafeMetadata,
      obj.emailAddresses,
      obj.phoneNumbers,
      obj.web3Wallets,
      obj.externalAccounts,
    );
  }
}
