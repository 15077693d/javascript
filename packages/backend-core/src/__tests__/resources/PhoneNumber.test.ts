import { ObjectType, PhoneNumber, PhoneNumberJSON } from '../../api/resources';

const mockPhoneNumberDTO: PhoneNumberJSON = {
  object: ObjectType.PhoneNumber,
  id: 'mock_phone_number_id',
  phone_number: '+12345678',
  reserved_for_second_factor: true,
  default_second_factor: true,
  linked_to: [],
  verification: null,
};

test('phone number defaults', function () {
  const phoneNumber = PhoneNumber.fromJSON(mockPhoneNumberDTO);
  expect(phoneNumber).toMatchSnapshot();
});
