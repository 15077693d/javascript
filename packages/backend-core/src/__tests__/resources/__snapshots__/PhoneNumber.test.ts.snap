// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`phone number defaults 1`] = `
PhoneNumber {
  "defaultSecondFactor": true,
  "id": "mock_phone_number_id",
  "linkedTo": Array [],
  "phoneNumber": "+12345678",
  "reservedForSecondFactor": true,
  "verification": null,
}
`;
